pvlib/pvarray.py

```diff
@@ -0,0 +1,219 @@
+       doi: 10.1109/JPHOTOV.2020.3045677.
+
+    Examples
+    --------
+    >>> pvefficiency_adr([1000, 200], 25,
+            k_a=100, k_d=-6.0, tc_d=0.02, k_rs=0.05, k_rsh=0.10)
+    array([100.        ,  92.79729308])
+
+    >>> pvefficiency_adr([1000, 200], 25,
+            k_a=1.0, k_d=-6.0, tc_d=0.02, k_rs=0.05, k_rsh=0.10)
+    array([1.        , 0.92797293])
+
+    '''
+    # Adapted from https://github.com/adriesse/pvpltools-python
+    # Copyright (c) 2022, Anton Driesse, PV Performance Labs
+    # All rights reserved.
+
+    k_a = np.array(k_a)
+    k_d = np.array(k_d)
+    tc_d = np.array(tc_d)
+    k_rs = np.array(k_rs)
+    k_rsh = np.array(k_rsh)
+
+    # normalize the irradiance
+    G_REF = np.array(1000.)
+    s = irradiance / G_REF
+
+    # obtain the difference from reference temperature
+    T_REF = np.array(25.)
+    dt = temperature - T_REF
+
+    # equation 29 in JPV
+    s_o     = exp10(k_d + (dt * tc_d))                             # noQA: E221
+    s_o_ref = exp10(k_d)
+
+    # equation 28 and 30 in JPV
+    # the constant k_v does not appear here because it cancels out
+    v  = np.log(s / s_o     + 1)                                   # noQA: E221
+    v /= np.log(1 / s_o_ref + 1)
+
+    # equation 25 in JPV
+    eta = k_a * ((1 + k_rs + k_rsh) * v - k_rs * s - k_rsh * v**2)
+
+    return eta
+
+
+def fit_pvefficiency_adr(irradiance, temperature, eta, dict_output=True,
+                         **kwargs):
+    """
+    Determine the parameters of the ADR module efficiency model by non-linear
+    least-squares fit to lab or field measurements.
+
+    Parameters
+    ----------
+    irradiance : numeric, non-negative
+        Effective irradiance incident on the PV module. [W/m²]
+
+    temperature : numeric
+        PV module operating temperature. [°C]
+
+    eta : numeric
+        Efficiency of the PV module at the specified irradiance and
+        temperature(s). [unitless] or [%]
+
+    dict_output : boolean, optional
+        When True (default), return the result as a dictionary; when False,
+        return the result as a numpy array.
+
+    kwargs :
+        Optional keyword arguments passed to `scipy.optimize.curve_fit`.
+        These kwargs can over-ride some options set within this function,
+        which could be interesting for very advanced users.
+
+    Returns
+    -------
+    popt : array or dict
+        Optimal values for the parameters.
+
+    Notes
+    -----
+    The best fits are obtained when the lab or field data include a wide range
+    of both irradiance and temperature values.  A minimal data set
+    would consist of 6 operating points covering low, medium and high
+    irradiance levels at two operating temperatures.
+
+    See also
+    --------
+    pvlib.pvarray.pvefficiency_adr
+    scipy.optimize.curve_fit
+
+    Adapted from https://github.com/adriesse/pvpltools-python
+    Copyright (c) 2022, Anton Driesse, PV Performance Labs
+    All rights reserved.
```

Comment:
Mild preference to put this in a comment in the function body as was done for `pvefficiency_adr`

Reply:
These lines were primarily intended to provoke discussion.  I will revert to my usual code comment.